inc/class-wp-sitemaps-renderer.php

```diff
@@ -195,15 +198,71 @@ public function get_sitemap_xml( $url_list ) {
 			)
 		);
 
+		$urlset->getNamespaces();
+
+		$attributes = array();
+
+		/**
+		 * Filters the `<urlset>` attributes for the sitemap.
+		 *
+		 * Can be used to add support for additional namespaces.
+		 *
+		 * @since 5.5.0
+		 *
+		 * @param array $attributes Associative array of urlset attributes and their values.
+		 */
+		$attributes = apply_filters( 'wp_sitemaps_urlset_attributes', $attributes );
+
+		foreach ( $attributes as $attribute => $value ) {
+			$urlset->addAttribute( 'xmlns:' . $attribute, $value );
+		}
+
 		foreach ( $url_list as $url_item ) {
 			$url = $urlset->addChild( 'url' );
 
-			// Add each attribute as a child node to the URL entry.
+			// Add each attribute as a child node to the <url> entry.
 			foreach ( $url_item as $attr => $value ) {
-				if ( 'url' === $attr ) {
-					$url->addChild( $attr, esc_url( $value ) );
+				$prefix = '';
+				if ( false !== strpos( $attr, ':' ) ) {
+					$prefix = explode( ':', $attr )[0] . ':';
+				}
+
+				/*
+				 * Arrays need to be prefixed with their namespace.
+				 *
+				 * Turns
+				 *
+				 * 'image:image' => array(
+				 *   array(
+				 *     'image:loc' => 'http://example.com/image.jpg',
+				 *     'image:title' => 'Cats',
+				 *   ),
+				 *   array(
+				 *     'image:loc' => 'http://example.com/image2.jpg',
+				 *     'image:title' => 'Cats and Dogs',
+				 *   ),
+				 * )
+				 *
+				 * into
+				 *
+				 * <image:image>
+				 *   <image:loc>http://example.com/image.jpg</image:loc>
+				 *   <image:title>Cats</image:loc>
+				 * </image:image>
+				 * <image:image>
+				 *   <image:loc>http://example.com/image2.jpg</image:loc>
+				 *   <image:title>Cats and Dogs</image:loc>
+				 * </image:image>
+				 */
+				if ( is_array( $value ) ) {
+					foreach ( $value as $child_attr => $child_value ) {
+						$item = $url->addChild( $prefix . $attr );
+						foreach ( (array) $child_value as $grandchild_attr => $grandchild_value ) {
```

Comment:
This is incorrect, since when `$child_value` is a string, `$grandchild_attr` will be `0` and an element with name `0` will be added :-(

Like the helper method mentioned below, it would be easier to add another helper method such as:

```
protected function add_child( $parent_element, $name, $value, $namespace_bindings ) {
	list( $prefix, $local_name ) = $this->expand_qname( $name );

	if ( is_array( $value ) ) {
		$child = $parent_element->addChild( $local_name, '', $namespace_bindings[ $prefix ] );
		foreach ( $value as $child_name => $child_value ) {
			list( $child_prefix, $child_local_name ) = $this->expand_qname( $child_name );

			$this->add_child( $child, $child_local_name, $child_value, $namespace_bindings[ $child_prefix ] );
		}
	} else {
		$parent_element->addChild( $local_name, esc_attr( $value ), $namespace_bindings[ $prefix ] );
	}
}
```

which would allow `grand_grand_children`, etc to be added.

And then the loop would simply become:

```
foreach ( $url_item as $attr => $value ) {
	$this->add_child( $url, $attr, $value );
}
```

Reply:
just a minor point, but using `$attr` there has always bugged me :-)

Coming at this from an XML perspective, I expect `$attr` to be used for "XML attributes" and this is an element name; hence, it would be easier to read if it were

```
foreach ( $url_item as $name=> $value ) {
	$this->add_child( $url, $name, $value );
}
```

Reply:
When would there ever be a need for grand grand children?

Reply:
the "need" is determined by whoever defines the markup for the extension.

Since the Sitemaps spec does **not** place an arbitrary max depth on elements in extension namespaces, I don't think core should either, especially since using a recursive helper method (as above) is trivially easy.  Thus, it seems to me that if an arbitrary max depth is to be enforced it should be up to those who want that enforced to argue for why _that_ is necessary.

Am I aware of any published specs for sitemaps extensions that currently require it?  No.

I know Video Sitemaps are out-of-scope **for core**, but consider the following.

Google's description of [Video Sitemaps](https://support.google.com/webmasters/answer/80471?hl=en&ref_topic=4581190) mentions that Google also supports [Media RSS](https://www.rssboard.org/media-rss).  One of the elements defined in Media RSS is [media:location](https://www.rssboard.org/media-rss#media-peerlink).  The documentation for that element says

> Optional element to specify geographical information about various locations captured in the content of a media object.  The format conforms to [geoRSS](http://www.georss.org/).

and then gives the example:

```
<media:location description="My house" start="00:01" end="01:00">
  <georss:where>
    <gml:Point>
      <gml:pos>35.669998 139.770004</gml:pos>
    </gml:Point>
  </georss:where>
</media:location>
```

Granted, there is no equivalent to `<media:location>` in the [published XML Schema](https://www.google.com/schemas/sitemap-video/1.1/sitemap-video.xsd) for Video Sitemaps (and that XML Schema does not allow any extensions), but if Google does **fully** support Media RSS for indexing videos, then it's possible that at some point in the future they will release a version of Video Sitemaps that also supports it...and _that_ would require grand grand children.

Again, I know Video Sitemaps are out-of-scope **for core**, but the whole point of #151 is to allow plugins to support them (in their current and **future** form) without needing to completely duplicate the infrastructure provided by core sitemaps...as long as providing that ability to plugins is not hard and I don't think it is.